### PR TITLE
Ingesting Loom Files into Firestore

### PR DESCRIPTION
**Pull request by knapii-developments**
This PR deals with [SCP-1831](https://broadworkbench.atlassian.net/browse/SCP-1831).  Only gene expression data is extracted and loaded into FireStore. It should be noted that row and column attributes,  or metadata, are not extracted from the file. 

Command:
python ingest_pipeline.py  --study-accession SCP1 --file-id 123abc ingest_expression --matrix-file ../tests/data/test_loom.loom  --matrix-file-type loom --taxon-name 'Homo Sapiens' --taxon-common-name humans

### CODE REVIEW

File: ingest/ingest_pipeline.py [hunk 1]
```diff
@@ -21,6 +21,9 @@
 # Ingest dense file
 python ingest_pipeline.py  --study-accession SCP1 --file-id 123abc ingest_expression --taxon-name 'Homo Sapiens' --taxon-common-name humans --matrix-file ../tests/data/dense_matrix_19_genes_100k_cells.txt --matrix-file-type dense
 
+# Ingest loom file
+python ingest_pipeline.py  --study-accession SCP1 --file-id 123abc ingest_expression --matrix-file ../tests/data/test_loom.loom  --matrix-file-type loom --taxon-name 'Homo Sapiens' --taxon-common-name humans
```
Comment (eweitz) [inline ingest/ingest_pipeline.py]:
```suggestion
python ingest_pipeline.py  --study-accession SCP1 --file-id 123abc ingest_expression --matrix-file ../tests/data/test_loom.loom  --matrix-file-type loom --taxon-name 'Homo sapiens' --taxon-common-name human
```

File: ingest/loom.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
+"""Module for ingesting Loom files into Firestore
+
+DESCRIPTION
+Module provides extract and transforms function for gene expression data for
+a loom File.
```
Comment (eweitz) [inline ingest/loom.py]:
```suggestion
a Loom file.
```

File: ingest/loom.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+        Returns:
+        ------
+            view: < generator LoomView>
+                    A generator of a view corresponding to the current chunk
+        """
+        for (ix, selection, view) in self.ds.scan(axis=0, batch_size=10):
+            yield view
+
+    def transform_expression_data_by_gene(self, view) -> Gene:
```
Comment (eweitz) [inline ingest/loom.py]: Would this type annotation be more correct?

```suggestion
    def transform_expression_data_by_gene(self, view) -> List[Gene]:
```

File: tests/mock_data/loom/gene_model_0.txt [hunk 4]
```diff
@@ -0,0 +1 @@
+{'name': 'A2M', 'gene_id': None, 'source_file_type': ('loom',), 'cell_names': ['BM01_16dpp_AAGCAGTGGTAT', 'BM01_16dpp_TAAGCAGTGGTA', 'BM01_16dpp_CTAAGCAGTGGT', 'BM01_16dpp_CGGTAAACCATT', 'BM01_16dpp_CCGAATTCACCG'], 'expression_scores': [-0.191, -0.191, 2.646, -0.191, -0.191], 'subdocument': {'cell_names': ['BM01_16dpp_AAGCAGTGGTAT', 'BM01_16dpp_TAAGCAGTGGTA', 'BM01_16dpp_CTAAGCAGTGGT', 'BM01_16dpp_CGGTAAACCATT', 'BM01_16dpp_CCGAATTCACCG'], 'expression_scores': [-0.191, -0.191, 2.646, -0.191, -0.191], 'source_file_type': 'loom'}, 'top_level_doc': {'file_id': '1234abc', 'searchable_name': 'a2m', 'name': 'A2M', 'gene_id': None, 'study_accession': 'SCP1', 'taxon_name': 'Homo Sapiens', 'taxon_common_name': 'human', 'ncbi_taxid': '9606', 'genome_assembly_accession': 'GCA_000001405.15', 'genome_annotation': 'Ensemble 94'}}
```
Comment (eweitz) [inline tests/mock_data/loom/gene_model_0.txt]: Fixing two minor typos (also updated in test):
* Homo Sapiens -> Homo sapiens
* Ensemble 94 -> Ensembl 94

```suggestion
{'name': 'A2M', 'gene_id': None, 'source_file_type': ('loom',), 'cell_names': ['BM01_16dpp_AAGCAGTGGTAT', 'BM01_16dpp_TAAGCAGTGGTA', 'BM01_16dpp_CTAAGCAGTGGT', 'BM01_16dpp_CGGTAAACCATT', 'BM01_16dpp_CCGAATTCACCG'], 'expression_scores': [-0.191, -0.191, 2.646, -0.191, -0.191], 'subdocument': {'cell_names': ['BM01_16dpp_AAGCAGTGGTAT', 'BM01_16dpp_TAAGCAGTGGTA', 'BM01_16dpp_CTAAGCAGTGGT', 'BM01_16dpp_CGGTAAACCATT', 'BM01_16dpp_CCGAATTCACCG'], 'expression_scores': [-0.191, -0.191, 2.646, -0.191, -0.191], 'source_file_type': 'loom'}, 'top_level_doc': {'file_id': '1234abc', 'searchable_name': 'a2m', 'name': 'A2M', 'gene_id': None, 'study_accession': 'SCP1', 'taxon_name': 'Homo sapiens', 'taxon_common_name': 'human', 'ncbi_taxid': '9606', 'genome_assembly_accession': 'GCA_000001405.15', 'genome_annotation': 'Ensembl 94'}}
```

File: tests/test_ingest.py [hunk 5]
```diff
@@ -318,6 +318,48 @@ def test_mtx_bundle_argument_validation(self):
...
+        """
+
+        args = [
+            "--study-accession",
+            "SCP1",
+            "--file-id",
+            "1234abc",
+            "ingest_expression",
+            "--taxon-name",
+            "Homo Sapiens",
```
Comment (eweitz) [inline tests/test_ingest.py]: Also updated in mock data.

```suggestion
            "Homo sapiens",
```

File: tests/test_ingest.py [hunk 6]
```diff
@@ -318,6 +318,48 @@ def test_mtx_bundle_argument_validation(self):
...
+            "--taxon-common-name",
+            "human",
+            "--ncbi-taxid",
+            "9606",
+            "--genome-assembly-accession",
+            "GCA_000001405.15",
+            "--genome-annotation",
+            "Ensemble 94",
```
Comment (eweitz) [inline tests/test_ingest.py]: Also updated in mock data.

```suggestion
            "Ensembl 94",
```

File: tests/test_ingest.py [hunk 7]
```diff
@@ -318,6 +318,48 @@ def test_mtx_bundle_argument_validation(self):
...
+            "../tests/data/test_loom.loom",
+            "--matrix-file-type",
+            "loom",
+        ]
+
+        ingest = self.setup_ingest(args)
+
+        models = ingest.load_expression_data_args[0]
+
+        # Verify that 25 gene models were passed into load method
```
Comment (eweitz) [inline tests/test_ingest.py]:
```suggestion
        # Verify that 10 gene models were passed into load method
```

File: ingest/ingest_pipeline.py [hunk 8]
```diff
@@ -413,6 +424,7 @@ def main() -> None:
     parsed_args = create_parser().parse_args()
     validate_arguments(parsed_args)
     arguments = vars(parsed_args)
+    print(arguments)
```
Comment (bistline) [inline ingest/ingest_pipeline.py]: Not a blocking comment, but let's try to be more mindful about these kind of debugging statements.  Eventually these will get written to Stackdriver (or some equivalent service) and we're going to want to be more consistent in how/what we're logging.

File: ingest/loom.py [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+        # TODO: Find way to utilize generators and interators for memory and CPU efficiency
+        # Checkout https://www.datacamp.com/community/tutorials/python-iterator-tutorial
+        gene_expression_models = []
+        for gene in view.ra.Gene:
+            gene_expression_models.append(
+                Gene(
+                    name=gene,
+                    source_file_type="loom",
+                    expression_scores=view[view.ra.Gene == gene, :][0],
+                    cell_names=self.ds.ca.CellID.tolist(),
```
Comment (bistline) [inline ingest/loom.py]: For future thought - we may want to include something in our file validations that if the file type is Loom that this particular metadata column (CellID) is present, and that we're checking for it in a case-insensitive fashion.

File: ingest/ingest_pipeline.py (hunk 1 above)
Comment (knapii-developments) [reply ingest/ingest_pipeline.py]: I actually performed a git mv.... But I will make sure to do that!

Comment (knapii-developments) [reply ingest/ingest_pipeline.py]: Addressed in f163441

File: ingest/loom.py (hunk 2 above)
Comment (knapii-developments) [reply ingest/loom.py]: Addressed in f708656.

File: ingest/loom.py (hunk 3 above)
Comment (knapii-developments) [reply ingest/loom.py]: Addressed in f708656

File: tests/mock_data/loom/gene_model_0.txt (hunk 4 above)
Comment (knapii-developments) [reply tests/mock_data/loom/gene_model_0.txt]: Addressed in f708656

File: tests/test_ingest.py (hunk 5 above)
Comment (knapii-developments) [reply tests/test_ingest.py]: Addressed in f708656

File: tests/test_ingest.py (hunk 6 above)
Comment (knapii-developments) [reply tests/test_ingest.py]: Addressed in f708656

File: tests/test_ingest.py (hunk 7 above)
Comment (knapii-developments) [reply tests/test_ingest.py]: Addressed in f708656

File: ingest/ingest_pipeline.py (hunk 8 above)
Comment (knapii-developments) [reply ingest/ingest_pipeline.py]: Addressed in f708656

**Pull request merged**